pkg/prometheus/promcfg.go

```diff
@@ -2078,3 +2082,134 @@ func (cg *ConfigGenerator) GenerateAgentConfiguration(
+	slices []yaml.MapSlice,
+	scrapeConfigs map[string]*v1alpha1.ScrapeConfig,
+	apiserverConfig *v1.APIServerConfig,
+	store *assets.Store,
+	shards int32) []yaml.MapSlice {
+	scrapeConfigIdentifiers := make([]string, len(scrapeConfigs))
+	i := 0
+	for k := range scrapeConfigs {
+		scrapeConfigIdentifiers[i] = k
+		i++
+	}
+
+	// Sorting ensures, that we always generate the config in the same order.
+	sort.Strings(scrapeConfigIdentifiers)
+
+	for _, identifier := range scrapeConfigIdentifiers {
+		slices = append(slices,
+			cg.WithKeyVals("scrapeconfig", identifier).generateScrapeConfig(
+				scrapeConfigs[identifier],
+				apiserverConfig,
+				store,
+				shards,
+			),
+		)
+	}
+
+	return slices
+}
+
+func (cg *ConfigGenerator) generateScrapeConfig(
+	sc *v1alpha1.ScrapeConfig,
+	apiserverConfig *v1.APIServerConfig,
+	store *assets.Store,
+	shards int32,
+) yaml.MapSlice {
+	jobName := fmt.Sprintf("scrapeconfig/%s/%s", sc.Namespace, sc.Name)
+	cfg := yaml.MapSlice{
+		{
+			Key:   "job_name",
+			Value: jobName,
+		},
+	}
+
+	cpf := cg.prom.GetCommonPrometheusFields()
+	labeler := namespacelabeler.New(cpf.EnforcedNamespaceLabel, cpf.ExcludedFromEnforcement, false)
+
+	if sc.Spec.HonorTimestamps != nil {
+		cfg = cg.AddHonorTimestamps(cfg, sc.Spec.HonorTimestamps)
+	}
+
+	if sc.Spec.HonorLabels != nil {
+		cfg = cg.AddHonorLabels(cfg, *sc.Spec.HonorLabels)
+	}
+
+	if sc.Spec.MetricsPath != "" {
+		cfg = append(cfg, yaml.MapItem{Key: "metrics_path", Value: sc.Spec.MetricsPath})
+	}
+
+	if sc.Spec.RelabelConfigs != nil {
+		cfg = append(cfg, yaml.MapItem{
+			Key:   "relabel_configs",
+			Value: labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs),
+		})
+	}
+
+	// StaticConfig
+	if len(sc.Spec.StaticConfigs) > 0 {
+		configs := make([][]yaml.MapItem, len(sc.Spec.StaticConfigs))
+		for i, config := range sc.Spec.StaticConfigs {
+			configs[i] = []yaml.MapItem{
+				{
+					Key:   "targets",
+					Value: config.Targets,
+				},
+				{
+					Key:   "labels",
+					Value: config.Labels,
+				},
+			}
+		}
+		cfg = append(cfg, yaml.MapItem{
+			Key:   "static_configs",
+			Value: configs,
+		})
+	}
+
+	// FileSDConfig
+	if len(sc.Spec.FileSDConfigs) > 0 {
+		configs := make([][]yaml.MapItem, len(sc.Spec.FileSDConfigs))
+		for i, config := range sc.Spec.FileSDConfigs {
+			configs[i] = []yaml.MapItem{
+				{
+					Key:   "files",
+					Value: config.Files,
+				},
+				{
+					Key:   "refresh_interval",
+					Value: config.RefreshInterval,
+				},
+			}
+		}
+		cfg = append(cfg, yaml.MapItem{
+			Key:   "file_sd_configs",
+			Value: configs,
+		})
+	}
+
+	// HTTPSDConfig
+	if len(sc.Spec.HTTPSDConfigs) > 0 {
+		configs := make([][]yaml.MapItem, len(sc.Spec.HTTPSDConfigs))
+		for i, config := range sc.Spec.HTTPSDConfigs {
+			configs[i] = []yaml.MapItem{
+				{
+					Key:   "url",
+					Value: config.URL,
+				},
+				{
+					Key:   "refresh_interval",
+					Value: config.RefreshInterval,
```

Comment:
need to check for nil otherwise Prometheus crashes :)

Reply:
25aa19594